qiskit/extensions/hamiltonian_gate.py

```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Gate described by the time evolution of a Hermitian Hamiltonian operator.
+"""
+
+from numbers import Number
+import numpy
+import scipy.linalg
+
+from qiskit.circuit import Gate, ParameterExpression
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit import QuantumRegister
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+from qiskit.extensions.exceptions import ExtensionError
+
+from .unitary import UnitaryGate
+
+
+class HamiltonianGate(Gate):
+    """Class for representing evolution by a Hermitian Hamiltonian operator as a gate. This gate
+    resolves to a UnitaryGate, which can be decomposed into basis gates if it is 2 qubits or
+    less, or simulated directly in Aer for more qubits. """
+
+    def __init__(self, data, time, label=None):
+        """Create a gate from a hamiltonian operator and evolution time parameter t
+
+        Args:
+            data (matrix or Operator): unitary operator.
+            time (float or complex): time evolution parameter.
+            label (str): unitary name for backend [Default: None].
+
+        Raises:
+            ExtensionError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, 'to_matrix'):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
```

Comment:
Would it be safer to check for types instead of attributes here? The `to_matrix` is probably safe both ways but `to_operator` might also be the attribute of another class than `BaseOperator`.

Reply:
I was trying to just replicate the API for UnitaryGate. Do you think it doesn't make sense here?

Reply:
To operator lets you initialize with a quantum channel object in the case where that channel is really a unitary channel. This might be a not-needed edge case for this gate.

Reply:
Ok, removing.

Reply:
Actually - I need something in here to be able to pass in Operators anyway, so is it a problem to leave it as is to handle the broader set of BaseOperators, or would it be better to change it to an isinstance(Operator)?